### PR TITLE
WebGLRenderer: Add `WEBGL_clip_cull_distance` support.

### PR DESCRIPTION
**Pull request by RenaudRohlinger**
[WEBGL_clip_cull_distance](https://registry.khronos.org/webgl/extensions/WEBGL_clip_cull_distance/) is now supported by default on the latest version of Chrome and Safari.

This PR adds support for the new `WEBGL_clip_cull_distance` extension. Made an example to demonstrate how to use the extension. Added the build for testing purpose. Will remove it once it's approved.

Here the live demo. Also in this example I'm tricking the renderer by enabling `isPoints = true` on the mesh and not specifying the `gl_PointSize` in the shader which seems to  produce a cool bug, I think it looks cool so I would like to let it like that:
https://raw.githack.com/mrdoob/three.js/dev/examples/webgl2_clipculldistance.html
[image](https://raw.githack.com/renaudrohlinger/three.js/culldistance/examples/webgl_clipculldistance.html)

### CODE REVIEW

**Comment by Mugen87** [comment]
I'm new to `WEBGL_clip_cull_distance`^^. 

When I understand the extension correctly, it allows clipping in the vertex shader which means it is an alternative to `three.js`'s clipping planes, right? 

Do you mind explaining why `gl_ClipDistance` is an array? What does the distance values represent?

> Also in this example I'm tricking the renderer by enabling isPoints = true on the mesh and not specifying the gl_PointSize in the shader which seems to produce a cool bug,

Even if this works would you mind not doing this 😇 ? Setting the read-only `is*` flags is not something that we want to promote in the examples.

File: src/renderers/webgl/WebGLProgram.js [hunk 1]
```diff
@@ -451,6 +462,8 @@ function WebGLProgram( renderer, cacheKey, parameters, bindingStates ) {
 
 	const customExtensions = parameters.isWebGL2 ? '' : generateExtensions( parameters );
 
+	const customWebGLVertexExtensions = generateVertexExtensions( parameters );
```
Comment (Mugen87) [inline src/renderers/webgl/WebGLProgram.js:465]: I think you can call this just `customVertexExtensions`.

**Comment by RenaudRohlinger** [comment]
> Do you mind explaining why gl_ClipDistance is an array? What does the distance values represent?

It's an array to handle more complex culling I believe, for example multiple culling shapes or handling a view frustum culling directly in the vertex shader. For example:
```glsl
// Define frustum planes in clip space
// Near and far planes
gl_ClipDistance[0] = gl_Position.z + 1.0; // near plane z = -1
gl_ClipDistance[1] = -gl_Position.z + 10.0; // far plane z = 10

// Left and right planes
gl_ClipDistance[2] = gl_Position.x + 5.0; // left plane x = -5
gl_ClipDistance[3] = -gl_Position.x + 5.0; // right plane x = 5

// Bottom and top planes
gl_ClipDistance[4] = gl_Position.y + 5.0; // bottom plane y = -5
gl_ClipDistance[5] = -gl_Position.y + 5.0; // top plane y = 5
```

Also there is `gl_CullDistance` available if we want to cull completely the primitive instead of just clipping the vertices of it. The documentation available indeed doesn't make things so it's mostly my understanding of the current implementation here 😄

File: src/renderers/webgl/WebGLProgram.js [hunk 2]
```diff
@@ -147,6 +147,17 @@ function generateExtensions( parameters ) {
 
 }
 
+function generateVertexExtensions( parameters ) {
+
+	const chunks = [
+		parameters.extensionClipCullDistance ? '#extension GL_ANGLE_clip_cull_distance : require' : '',
+		parameters.extensionClipCullDistance ? '#extension GL_EXT_clip_cull_distance : enable' : '',
```
Comment (lexaknyazev) [inline src/renderers/webgl/WebGLProgram.js:154]: Only the `ANGLE_` string is needed.

**Comment by lexaknyazev** [comment]
> Do you mind explaining why `gl_ClipDistance` is an array? What does the distance values represent?

Each vertex may have up to 8 clip distance values; they are interpolated across primitives as regular varyings. During primitive rasterization, pixels that have negative clip distance values are discarded. This should be more efficient than using `discard` in fragment shaders.

For example, here's how you can turn a unit quad into an octagon using only a vertex shader:
```glsl
void main() {
    gl_Position = vec4(a_position, 0.0, 1.0);

    gl_ClipDistance[0] = dot(gl_Position, vec4( 1,  0, 0, 0.5));
    gl_ClipDistance[1] = dot(gl_Position, vec4(-1,  0, 0, 0.5));
    gl_ClipDistance[2] = dot(gl_Position, vec4( 0,  1, 0, 0.5));
    gl_ClipDistance[3] = dot(gl_Position, vec4( 0, -1, 0, 0.5));
    gl_ClipDistance[4] = dot(gl_Position, vec4( 1,  1, 0, 0.70710678));
    gl_ClipDistance[5] = dot(gl_Position, vec4( 1, -1, 0, 0.70710678));
    gl_ClipDistance[6] = dot(gl_Position, vec4(-1,  1, 0, 0.70710678));
    gl_ClipDistance[7] = dot(gl_Position, vec4(-1, -1, 0, 0.70710678));
}
```

**Comment by lexaknyazev** [comment]
Drawing points without specifying `gl_PointSize` is undefined and may produce inconsistent results on different platforms.

**Comment by RenaudRohlinger** [comment]
Updated with the feedbacks.

FYI since I'm using a builtin material I attacehd a custom object `extensions` in order to make the extension work since it's only available on `ShaderMaterial`:
```js
material = new THREE.MeshNormalMaterial();
material.extensions = {
	clipCullDistance: true,
};
```
Are you ok with this logic or do you have anything better in mind @Mugen87? For example we could move the `extensions` property from `ShaderMaterial` to `Material`?

File: examples/webgl_clipculldistance.html [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+        target="_blank"
+        rel="noopener"
+        >three.js</a
+      >
+      - clip cull distance
+      <a
+        href="https://registry.khronos.org/webgl/extensions/EXT_clip_cull_distance/"
+        target="_blank"
+        rel="noopener"
+        >EXT_clip_cull_distance</a
```
Comment (lexaknyazev) [inline examples/webgl_clipculldistance.html:29]: Please use the `WEBGL_` extension name and link.

**Comment by Mugen87** [comment]
> Are you ok with this logic or do you have anything better in mind @Mugen87? For example we could move the extensions property from ShaderMaterial to Material?

For now, let's make an example based on `ShaderMaterial`. We can think of moving `extensions` to a `Material` in a different PR/issue. Maybe there is also the possibility to integrate `WEBGL_clip_cull_distance` in the shader chunks and configure the distances with a new material property.

File: examples/webgl_clipculldistance.html (hunk 3 above)
Comment (Mugen87) [reply examples/webgl_clipculldistance.html]: Fixed!

**Comment by Mugen87** [comment]
@RenaudRohlinger I've updated the example to ensure formatting matches all other examples 👍 .

File: src/renderers/webgl/WebGLExtensions.js [hunk 4]
```diff
@@ -54,6 +54,7 @@ function WebGLExtensions( gl ) {
 			if ( capabilities.isWebGL2 ) {
 
 				getExtension( 'EXT_color_buffer_float' );
+				getExtension( 'WEBGL_clip_cull_distance' );
```
Comment (Mugen87) [inline src/renderers/webgl/WebGLExtensions.js:57]: Optional extensions are queried at the startup of the renderer to check if they are available. I've added `WEBGL_clip_cull_distance` to the WebGL 2 section.

File: src/renderers/webgl/WebGLPrograms.js [hunk 5]
```diff
@@ -356,6 +356,7 @@ function WebGLPrograms( renderer, cubemaps, cubeuvmaps, extensions, capabilities
 			extensionFragDepth: HAS_EXTENSIONS && material.extensions.fragDepth === true,
 			extensionDrawBuffers: HAS_EXTENSIONS && material.extensions.drawBuffers === true,
 			extensionShaderTextureLOD: HAS_EXTENSIONS && material.extensions.shaderTextureLOD === true,
+			extensionClipCullDistance: HAS_EXTENSIONS && material.extensions.clipCullDistance && extensions.has( 'WEBGL_clip_cull_distance' ),
```
Comment (Mugen87) [inline src/renderers/webgl/WebGLPrograms.js:359]: The changes to `WebGLProgram` and `WebGLPrograms` look good now!

File: examples/webgl_clipculldistance.html [hunk 6]
```diff
@@ -0,0 +1,249 @@
...
+				const width = window.innerWidth;
+				const height = window.innerHeight;
+
+				camera = new THREE.PerspectiveCamera( 70, width / height, 1, 100 );
+				camera.position.z = 30;
+
+				scene = new THREE.Scene();
+
+				renderer = new THREE.WebGLRenderer();
+				renderer.setPixelRatio( 1 );
```
Comment (gkjohnson) [inline examples/webgl_clipculldistance.html:70]: Why not set the pixel ratio to the device pixel ratio? The example looks low quality on my machine

Comment (Mugen87) [reply examples/webgl_clipculldistance.html:70]: I think we can also turn on anti-aliasing.

**Comment by gkjohnson** [comment]
Beyond potential performance benefits of performing plane clipping with this extension vs using discard in the fragment shader is that it enables a hardware anti-aliased edge on the clipped edges:

image

#22172 implemented AA on edges using alpha to coverage but this solution would be better, I think. I'd say this should be used to implement Material.clipPlanes but I don't know if `Material.clipIntersection === true` (ie unioned clipping planes) is possible with this technique.

**Comment by RenaudRohlinger** [comment]
> For now, let's make an example based on ShaderMaterial

Updated the material with `ShaderMaterial` /cc @Mugen87

**Comment by RenaudRohlinger** [comment]
@gkjohnson FYI  `WEBGL_clip_cull_distance ` is not yet supported, here to follow the current development:
[https://bugzilla.mozilla.org/show_bug.cgi?id=1854001](https://bugzilla.mozilla.org/show_bug.cgi?id=1854001)

**Comment by lexaknyazev** [comment]
Since clipping distances are set per-vertex, each primitive gets its own clipping half-space. For instance, this could be used to make it look like clipping happens against a "sphere".

**Comment by lexaknyazev** [comment]
> how we could integrate the extension in the material system

It could be used for HW acceleration of regular clipping planes, not explicitly exposed to users at all. Anything beyond that would likely require a custom shader.

> How common is it to define additional vertex data attributes (next to vertex normals, colors etc.) to configure clipping distances per vertex?

Never saw such usage. Clip distances are supposed to be dynamically computed akin to `gl_Position`.

**Comment by Mugen87** [comment]
> It could be used for HW acceleration of regular clipping planes, not explicitly exposed to users at all.

This is something that would could try to tackle when this PR is merged.

**Comment by Mugen87** [comment]
@RenaudRohlinger I've simplified the example so the focus lies more on the extension handling. It is not that visually appealing like your original draft but maintenance aspect is more important right now.

Do you mind reverting the changes to the build files and regenerate the screenshot? I think this PR can be merged then.

**Comment by RenaudRohlinger** [comment]
Awesome! LGTM, thanks @Mugen87

**Review by Mugen87 — APPROVED** [review]

**Pull request merged**